google/cloud/forseti/scanner/scanners/bigquery_scanner.py

```diff
@@ -115,32 +121,61 @@ def _retrieve(self):
 
         Returns:
             list: BigQuery ACL data
+
+        Raises:
+            ValueError: if resources have an unexpected type.
         """
         model_manager = self.service_config.model_manager
         scoped_session, data_access = model_manager.get(self.model_name)
         with scoped_session as session:
-            bq_acls = []
+            bq_acl_data = []
 
             for policy in data_access.scanner_iter(session, 'dataset_policy'):
-                dataset_id = policy.parent.name
+                # dataset_policy are always in a dataset, which is always in a
+                # project.
+                dataset = policy.parent
+                if dataset.type != 'dataset':
+                    raise ValueError(
+                        'Unexpected type of dataset_policy parent: '
+                        'got %s, want dataset' % dataset.type
+                    )
+
+                if dataset.parent.type != 'project':
+                    raise ValueError(
+                        'Unexpected type of dataset_policy grandparent: '
+                        'got %s, want project' % dataset.parent.type
+                    )
+
+                proj = project.Project(
+                    project_id=dataset.parent.name,
+                    full_name=dataset.parent.full_name,
+                    data=policy.data,
+                )
+
                 # There is no functional use for project_id in this scanner,
                 # other than to identify where the dataset comes from,
                 # which can now be done with full_name.
                 # In case you are tempted to get the project_id anyways,
                 # do not use project_id = policy.parent.parent.name
                 # which will cause db session conflict.
                 # Instead, parse the project_id from the full_name.
-                bq_acls.extend(
-                    BigqueryAccessControls.from_json(
-                        project_id=None,
-                        dataset_id=dataset_id,
-                        full_name=policy.full_name,
-                        acls=policy.data))
+                gen = BigqueryAccessControls.from_json(
```

Comment:
Simplified to converting to list + rename to  bq_acls.

I could replace project_id with parent_project, if that's what you mean? We need the parent Resouce class so just the id alone won't be enough.

The reason I made BigqueryAccessControlsData was to match other scanners. For example, all scanners (such as iam), return data which is usually a tuple of (parent_resource, policy) and then that is passed to find_violations. Instead of doing a tuple, I just made it a namedtuple. I feel that is easier to follow as I was getting a bit confused by the tuple in some parts. If you disagree I can make it a tuple like the rest.

Also, in bigquery_rules_engine.py, the BigqueryAccessControls object is also used as the rule itself. So while we could just omit the project for that, it may be cleaner to keep these two concepts separated.

What are your thoughts?

Reply:
Sounds good, thanks for the clarifications!